core/types/transaction_signing.go

```diff
@@ -534,20 +534,28 @@ type CachedSigner struct {
 	cache SenderCache
 }
 
-func WrapWithCachedSigner(signer Signer, cache SenderCache) *CachedSigner {
-	return &CachedSigner{
+func WrapWithCachedSigner(signer Signer, cache SenderCache) CachedSigner {
+	return CachedSigner{
 		Signer: signer,
 		cache:  cache,
 	}
 }
 
+
+func (cs CachedSigner) Equal(s2 Signer) bool {
+	cs2, ok := s2.(CachedSigner)
+	if ok {
+		// unwrap the signer
+		return cs.Signer.Equal(cs2.Signer)
```

Comment:
A problem we might get here is that this implementation of `Equal` is not symmetric. You could have a case like this:
```
s := FrontierSigner{}
cs := WrapWithCachedSigner(s, cache)
cs.Equal(s)  // == true
s.Equal(cs)  // == false
```
This is something that usually would not be expected of an `Equal` function, and it makes the effectiveness of the Cache dependent on the order of operands used in the comparison.

I tested the fix, and it seems to be used in a favorable way right now. But the implementation is brittle.

One way to fix the issue could be to remove the `Equal` from the `Signer` interface and replace it with an `Algorithm()`  method like this:
```
type Signer interface {
    // ... everything that is there now except Equal()
    Algorithm() SigningAlgorithm   // < all implementations return themselves except the cache one, which returns the algorithm of the wrapped signer
}

type SigningAlgorithm interface {
    Equals(SigningAlgorithm) bool  // each algorithm is only equal to itself, as it is now for all non-cached signers
}
```

Reply:
There are only two cases that comparing CachedSigner with (1) another CachedSigner; or (2) a non-cached Signer.

https://github.com/Fantom-foundation/go-ethereum/blob/634958c14f331bd58da8c19243f848fec1269675/core/types/transaction_signing.go#L556
cache.Get(tx.Hash()) could return a Signer or a CachedSigner.

https://github.com/Fantom-foundation/go-ethereum/blob/634958c14f331bd58da8c19243f848fec1269675/core/types/transaction_signing.go#L557
will then compare either a Signer-to-a-Signer, or a CachedSigner-to-a-Signer.
Both cases are handled already.

Regarding the suggestion to compare signers using SigningAlgorithm. It's interesting, but more changes will be need. But what if there were two cached signers (or signers) of the same Algorithm, would these signers actually be equal all the time? I guess not.

If more unit tests are needed, then please add them in.

Reply:
An example of a problematic call is here:
https://github.com/Fantom-foundation/go-ethereum/blob/634958c14f331bd58da8c19243f848fec1269675/core/types/transaction_signing.go#L137

In this line, the types of `sigCache.signer` and `signer` is the `Signer` interface. Thus, `sigCache.signer` is not restricted to `CachedSigner`. The "Cache" in the name `sigCache` refers to a cached sender address [here](https://github.com/Fantom-foundation/go-ethereum/blob/634958c14f331bd58da8c19243f848fec1269675/core/types/transaction_signing.go#L34). 

In such a context, it would be rather surprising if 
```
sigCache.signer.Equal(signer)
```
and
```
signer.Equal(sigCache.signer)
```
can lead to different results. Wouldn't you agree?

This is what should be avoided by making sure `Equal` is symmetric.

Reply:
Well, that function would assume param signer is a non-cached Signer. If it's not clear enough, perhaps a new check should be added.

Whilst agreeing with you about Equal should be symmetric (in general), the suggestion may need more changes.
Simply comparing using Algorithm may not be sufficient. See:
https://github.com/Fantom-foundation/go-ethereum/blob/634958c14f331bd58da8c19243f848fec1269675/core/types/transaction_signing.go#L198
https://github.com/Fantom-foundation/go-ethereum/blob/634958c14f331bd58da8c19243f848fec1269675/core/types/transaction_signing.go#L352

Reply:
Since this version was the one we tested, I suggest to go with this one for now. Then, indeed we can simply return an ID of the signer (`Equal` method is only used in this file, so no compatibility issues). Or, force symmetry by double checking:

```
func (s signer) Equal(s2 Signer) bool {
	return ..... || s2.Equal(s)
}
```
in each `Equal` method - it's ugly but doesn't change the interface